packages/frontend-test-utils/src/app/createExtensionTester.tsx

```diff
@@ -0,0 +1,299 @@
+    throw new Error('AppRouter must be rendered within the AppProvider');
+  }
+
+  const { appIdentityProxy } = internalAppContext;
+  const [identityApi, setIdentityApi] = useState<IdentityApi>();
+
+  if (!SignInPage) {
+    appIdentityProxy.setTarget(
+      {
+        getUserId: () => 'guest',
+        getIdToken: async () => undefined,
+        getProfile: () => ({
+          email: 'guest@example.com',
+          displayName: 'Guest',
+        }),
+        getProfileInfo: async () => ({
+          email: 'guest@example.com',
+          displayName: 'Guest',
+        }),
+        getBackstageIdentity: async () => ({
+          type: 'user',
+          userEntityRef: 'user:default/guest',
+          ownershipEntityRefs: ['user:default/guest'],
+        }),
+        getCredentials: async () => ({}),
+        signOut: async () => {},
+      },
+      { signOutTargetUrl },
+    );
+
+    return children;
+  }
+
+  if (!identityApi) {
+    return <SignInPage onSignInSuccess={setIdentityApi} />;
+  }
+
+  appIdentityProxy.setTarget(identityApi, {
+    signOutTargetUrl,
+  });
+
+  return children;
+};
+
+const TestCoreRouterExtension = createExtension({
+  namespace: 'core',
+  name: 'router',
+  attachTo: { id: 'core', input: 'root' },
+  inputs: {
+    signInPage: createExtensionInput(
+      {
+        component: createSignInPageExtension.componentDataRef,
+      },
+      { singleton: true, optional: true },
+    ),
+    children: createExtensionInput(
+      {
+        element: coreExtensionData.reactElement,
+      },
+      { singleton: true },
+    ),
+  },
+  output: {
+    element: coreExtensionData.reactElement,
+  },
+  factory({ inputs }) {
+    const SignInPage = inputs.signInPage?.output.component;
+    const children = inputs.children.output.element;
+
+    return {
+      element: (
+        <MemoryRouter>
+          <AuthenticationProvider signInPage={SignInPage}>
+            {children}
+          </AuthenticationProvider>
+        </MemoryRouter>
+      ),
+    };
+  },
+});
+
+/** @public */
+export class ExtensionTester {
+  /** @internal */
+  static forSubject<TConfig>(
+    subject: ExtensionDefinition<TConfig>,
+    options?: { config?: TConfig },
+  ): ExtensionTester {
+    const tester = new ExtensionTester();
+    const { output, factory, ...rest } = toInternalExtensionDefinition(subject);
+    // attaching to core/routes to render as index route
+    const extension = createExtension({
+      ...rest,
+      attachTo: { id: 'core/routes', input: 'routes' },
+      output: {
+        ...output,
+        path: coreExtensionData.routePath,
+      },
+      factory: params => ({
+        ...factory(params),
+        path: '/',
+      }),
+    });
+    tester.add(extension, options);
+    return tester;
+  }
+
+  readonly #extensions = new Array<{
+    id: string;
+    definition: ExtensionDefinition<any>;
+    config?: JsonValue;
+  }>();
+
+  add<TConfig>(
+    extension: ExtensionDefinition<TConfig>,
```

Comment:
not important but, I believe this could in theory just have been `ExtensionDefinition<any>` and skipping the template parameter right 🤔 feel free to skip

Reply:
We use it for the config option below